### PR TITLE
OvmfPkg/CryptoPkg: Enhance C runtime functions

### PR DESCRIPTION
**Pull request by leenasoman**
As part of the effort to get Boringssl to work with EDK2, there are many C runtime apis that get referenced and are currently not implemented in EDK2. This change adds those apis to edk2.

Cc: Ard Biesheuvel <ardb+tianocore@kernel.org>
    Cc: Jordan Justen <jordan.l.justen@intel.com>
    Cc: Laszlo Ersek <lersek@redhat.com>
    Cc: Anthony Perard <anthony.perard@citrix.com>
    Cc: Julien Grall <julien@xen.org>

# Description
In particular, this change adds the following :
1. Stubs for file functions. These have been implemented to return -1 rather than 0 since there is no file system support and to fail any calls that are call into them.

3. Implementations of the following api which are referenced and used by Boringssl code :
  - strdup : string duplication function
  -  bsearch : binary search
  -  getentropy : return entropy of requested length
  -  __ashlti3 : Left shift for 128 bit integers

- [ ] Breaking change?
  - **Breaking change** - Does this PR cause a break in build or boot behavior?
  - Examples: Does it add a new library class or move a module to a different repo.
- [ ] Impacts security?
  - **Security** - Does this PR have a direct security impact?
  - Examples: Crypto algorithm change or buffer overflow fix.
- [ ] Includes tests?
  - **Tests** - Does this PR include any explicit test code?
  - Examples: Unit tests or integration tests.

## How This Was Tested
Calls to the apis are successful when tested locally.
## Integration Instructions
N/A

### CODE REVIEW

**Comment by ardbiesheuvel** [comment]
Can you clean up the patch titles and explanations please? I see two patches with identical names.

**Comment by ardbiesheuvel** [comment]
please avoid merge commits - instead, rebase your changes and force-push them to the existing branch.

Also, I expect to see a logical sequence of patches here (if more than one is needed)

File: CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c [hunk 1]
```diff
@@ -580,15 +582,168 @@ fopen (
...
+
+  if (length > GETENTROPY_MAX) {
+    errno = EIO;
+    return -1;
+  }
+
+  if ((entropyBuffer == NULL) || (length == 0)) {
+    errno = EFAULT;
+    return -1;
+  }
```
Comment (deeglaze) [inline CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c:731]: from getentropy(3):

> EFAULT Part or all of the buffer specified by buffer and length
>               is not in valid addressable memory.

length == 0 is not a fault condition

File: CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c [hunk 2]
```diff
@@ -580,15 +582,168 @@ fopen (
...
+  for (i = 0; i < length; i += sizeof (UINT64)) {
+    if (!GetRandomNumber64 (&randNum)) {
+      errno = ENOSYS;
+      return -1;
+    }
+
+    copyLength =
+        (length - i >= sizeof (UINT64)) ? sizeof (UINT64) : (length - i);
+    CopyMem (entropyBuffer + i, &randNum, copyLength);
+  }
```
Comment (deeglaze) [inline CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c:742]:
```suggestion
  if (!length) {
    return 0;
  }
  if (!ArchIsRngSupported ()) {
    errno = ENOSYS;
    return -1;
  }
  if (entropyBuffer == NULL) {
    errno = EFAULT;
    return -1;
  }
  if (length > GETENTROPY_MAX) {
    errno = EIO;
    return -1;
  }
  copyLength = (length % sizeof(UINT64));
  if (copyLength) {
    ArchGetRandomNumber64(&randNum);
    CopyMem(entropyBuffer, &randNum, copyLength);
  }
  for (i = copyLength; i < length; i += sizeof (UINT64)) {
    ArchGetRandomNumber64 (entropyBuffer + i);
  }
  return 0;
}
```

There's no reason to put the conditionals in a loop other than the loop condition.

File: CryptoPkg/Library/Include/CrtLibSupport.h [hunk 3]
```diff
@@ -437,6 +472,66 @@ strcat (
...
+  char  *strSource
+  );
+
+void *
+bsearch (
+  const void *key,
+  const void *base,
+  size_t nitems,
+  size_t width,
+  int         ( *cmp )(const void *, const void *)
```
Comment (deeglaze) [inline CryptoPkg/Library/Include/CrtLibSupport.h:526]:
```suggestion
  size_t nmemb,
  size_t size,
  int       ( *compar )(const void *, const void *)
```

Use the same names as bsearch(3).

File: CryptoPkg/Library/Include/CrtLibSupport.h [hunk 4]
```diff
@@ -437,6 +472,66 @@ strcat (
...
+  );
+
+int
+ftell (
+  FILE  *fp
+  );
+
+int
+fputs (
+  const char  *str,
```
Comment (deeglaze) [inline CryptoPkg/Library/Include/CrtLibSupport.h:511]:
```suggestion
  const char  *restrict s,
```
Be compatible with fputs(3). I don't see other uses of `restrict` in EDK2, but I don't know why we wouldn't use it. Would need a maintainer to comment.

File: CryptoPkg/Library/Include/CrtLibSupport.h [hunk 5]
```diff
@@ -437,6 +472,66 @@ strcat (
...
+  FILE        *stream
+  );
+
+char *
+strdup (
+  char  *strSource
```
Comment (deeglaze) [inline CryptoPkg/Library/Include/CrtLibSupport.h:517]:
```suggestion
  char  *s
```
strdup(3) name choice.

A shame, but it is what it is.

File: CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c [hunk 6]
```diff
@@ -580,15 +582,168 @@ fopen (
...
+  long  offset,
+  int   whence
+  )
+{
+  return -1;
+}
+
+int
+feof (
+  FILE  *fp
```
Comment (deeglaze) [inline CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c:631]:
```suggestion
  FILE  *stream
```
feof(3) name.

File: CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c [hunk 7]
```diff
@@ -580,15 +582,168 @@ fopen (
...
+  )
+{
+  return -1;
+}
+
+int
+ftell (
+  FILE  *fp
```
Comment (deeglaze) [inline CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c:639]:
```suggestion
  FILE  *stream
```
ftell(3) name.

File: CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c [hunk 8]
```diff
@@ -580,15 +582,168 @@ fopen (
...
+  )
+{
+  return -1;
+}
+
+char *
+fgets (
+  char  *str,
+  int   n,
```
Comment (deeglaze) [inline CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c:648]:
```suggestion
  char  *s,
  int   size,
```
fgets(3) names.

File: CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c [hunk 9]
```diff
@@ -580,15 +582,168 @@ fopen (
...
+  FILE  *stream
+  )
+{
+  return NULL;
+}
+
+char *
+strdup (
+  char  *strSource
```
Comment (deeglaze) [inline CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c:657]:
```suggestion
  char  *s
```
strdup(3) name.

File: CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c [hunk 10]
```diff
@@ -580,15 +582,168 @@ fopen (
...
+  )
+{
+  UINTN  len;
```
Comment (deeglaze) [inline CryptoPkg/Library/BaseCryptLib/SysCall/CrtWrapper.c:660]:
```suggestion
  UINTN  Length;
```
Follow EDK2 style guide.

File: CryptoPkg/Library/Include/CrtLibSupport.h [hunk 11]
```diff
@@ -437,6 +472,66 @@ strcat (
...
   );
 
+int
+fflush (
+  FILE  *stream
+  );
+
+int
+ferror (
+  FILE  *fp
```
Comment (deeglaze) [inline CryptoPkg/Library/Include/CrtLibSupport.h:482]:
```suggestion
  FILE  *stream
```
ferror(3) name.

File: CryptoPkg/Library/Include/CrtLibSupport.h [hunk 12]
```diff
@@ -437,6 +472,66 @@ strcat (
...
+int
+fseek (
+  FILE  *stream,
+  long  offset,
+  int   whence
+  );
+
+int
+feof (
+  FILE  *fp
```
Comment (deeglaze) [inline CryptoPkg/Library/Include/CrtLibSupport.h:494]:
```suggestion
  FILE  *stream
```
feof(3) name.

File: CryptoPkg/Library/Include/CrtLibSupport.h [hunk 13]
```diff
@@ -437,6 +472,66 @@ strcat (
...
+  );
+
+char *
+fgets (
+  char  *str,
+  int   n,
```
Comment (deeglaze) [inline CryptoPkg/Library/Include/CrtLibSupport.h:500]:
```suggestion
  char  *s,
  int   size,
```
fgets(3) names.

File: CryptoPkg/Library/Include/CrtLibSupport.h [hunk 14]
```diff
@@ -437,6 +472,66 @@ strcat (
...
+  FILE  *stream
+  );
+
+int
+ftell (
+  FILE  *fp
```
Comment (deeglaze) [inline CryptoPkg/Library/Include/CrtLibSupport.h:506]:
```suggestion
  FILE  *stream
```
ftell(3) name.

**Review by deeglaze — CHANGES_REQUESTED** [review]

**Pull request closed without merging**